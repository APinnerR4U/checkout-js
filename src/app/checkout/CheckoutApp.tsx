import { createCheckoutService, createEmbeddedCheckoutMessenger } from '@bigcommerce/checkout-sdk';
import { createRequestSender } from '@bigcommerce/request-sender';
import React, { Component } from 'react';
import ReactModal from 'react-modal';

import { StepTrackerFactory } from '../analytics';
import { ErrorLoggerFactory, ErrorLoggingBoundary } from '../common/error';
import { NewsletterService } from '../customer';
import { createEmbeddedCheckoutStylesheet, createEmbeddedCheckoutSupport } from '../embeddedCheckout';
import { getLanguageService, LocaleProvider } from '../locale';
import { FlashMessage } from '../ui/alert';

import Checkout from './Checkout';
import CheckoutProvider from './CheckoutProvider';

export interface CheckoutAppProps {
    checkoutId: string;
    containerId: string;
    flashMessages: FlashMessage[]; // TODO: Expose flash messages from SDK
}

export default class CheckoutApp extends Component<CheckoutAppProps> {
    private checkoutService = createCheckoutService({
        locale: getLanguageService().getLocale(),
        shouldWarnMutation: process.env.NODE_ENV === 'development',
    });
    private embeddedSupport = createEmbeddedCheckoutSupport(getLanguageService());
    private embeddedStylesheet = createEmbeddedCheckoutStylesheet();
    private errorLogger = new ErrorLoggerFactory().getLogger();
    private newsletterService = new NewsletterService(createRequestSender());
    private stepTrackerFactory = new StepTrackerFactory(this.checkoutService);

    componentDidMount(): void {
        const { containerId } = this.props;

        ReactModal.setAppElement(`#${containerId}`);
    }

    render() {
        return (
            <ErrorLoggingBoundary logger={ this.errorLogger }>
                <LocaleProvider checkoutService={ this.checkoutService }>
                    <CheckoutProvider checkoutService={ this.checkoutService }>
                        <Checkout
                            { ...this.props }
                            createEmbeddedMessenger={ createEmbeddedCheckoutMessenger }
                            createStepTracker={ () => this.stepTrackerFactory.createTracker() }
                            embeddedStylesheet={ this.embeddedStylesheet }
                            embeddedSupport={ this.embeddedSupport }
                            errorLogger={ this.errorLogger }
                            subscribeToNewsletter={ data => this.newsletterService.subscribe(data) }
                        />
                    </CheckoutProvider>
                </LocaleProvider>
            </ErrorLoggingBoundary>
        );
    }
}
