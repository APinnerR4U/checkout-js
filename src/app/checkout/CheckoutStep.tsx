import classNames from 'classnames';
import { noop } from 'lodash';
import React, { createRef, Component, ReactNode, RefObject } from 'react';
import { CSSTransition } from 'react-transition-group';

import CheckoutStepHeader from './CheckoutStepHeader';
import CheckoutStepType from './CheckoutStepType';

export interface CheckoutStepProps {
    heading?: ReactNode;
    isActive?: boolean;
    isComplete?: boolean;
    isEditable?: boolean;
    summary?: ReactNode;
    type: CheckoutStepType;
    onExpanded?(step: CheckoutStepType): void;
    onEdit?(step: CheckoutStepType): void;
}

const LARGE_SCREEN_BREAKPOINT = 968;
const LARGE_SCREEN_ANIMATION_DELAY = 610;

export default class CheckoutStep extends Component<CheckoutStepProps> {
    private containerRef: RefObject<HTMLElement> = createRef();
    private timeoutRef?: number;

    componentDidMount(): void {
        const { isActive } = this.props;

        if (isActive) {
            this.focusStep();
        }
    }

    componentDidUpdate(prevProps: Readonly<CheckoutStepProps>): void {
        const { isActive } = this.props;

        if (isActive && isActive !== prevProps.isActive) {
            this.focusStep();
        }
    }

    componentWillUnmount(): void {
        if (this.timeoutRef) {
            window.clearTimeout(this.timeoutRef);

            this.timeoutRef = undefined;
        }
    }

    render(): ReactNode {
        const { children } = this.props;

        const {
            heading,
            isActive,
            isComplete,
            isEditable,
            onEdit,
            summary,
            type,
        } = this.props;

        return (
            <li
                className={ classNames(
                    'checkout-step',
                    'optimizedCheckout-checkoutStep',
                    { [`checkout-step--${type}`]: !!type }
                ) }
                ref={ this.containerRef as RefObject<HTMLLIElement> }
            >
                <div className="checkout-view-header">
                    <CheckoutStepHeader
                        heading={ heading }
                        isActive={ isActive }
                        isEditable={ isEditable }
                        isComplete={ isComplete }
                        onEdit={ onEdit }
                        summary={ summary }
                        type={ type }
                    />
                </div>

                <CSSTransition
                    addEndListener={ (node, done) => {
                        node.addEventListener('transitionend', ({ target }) => {
                            if (target === node) {
                                done();
                            }
                        });
                    } }
                    classNames="checkout-view-content"
                    timeout={ {} }
                    in={ isActive }
                    unmountOnExit
                    mountOnEnter
                >
                    <div className="checkout-view-content">
                        { children }
                    </div>
                </CSSTransition>
            </li>
        );
    }

    private focusStep(): void {
        const delay = window.innerWidth > LARGE_SCREEN_BREAKPOINT ? LARGE_SCREEN_ANIMATION_DELAY : 0;

        this.timeoutRef = window.setTimeout(() => {
            const input = this.getChildInput();
            const position = this.getScrollPosition();
            const { type, onExpanded = noop } = this.props;

            if (input) {
                input.focus();
            }

            if (position !== undefined && !isNaN(position)) {
                window.scrollTo(0, position);
            }

            onExpanded(type);

            this.timeoutRef = undefined;
        }, delay);
    }

    private getChildInput(): HTMLElement | undefined {
        const container = this.containerRef.current;

        if (!container) {
            return;
        }

        const input = container.querySelector<HTMLElement>('input, select, textarea');

        return input ? input : undefined;
    }

    private getScrollPosition(): number | undefined {
        const container = this.getParentContainer();
        const { isComplete } = this.props;

        if (!container || window !== window.top) {
            return;
        }

        const topOffset = isComplete ? 0 : window.innerHeight / 5;
        const containerOffset = container.getBoundingClientRect().top + (window.scrollY || window.pageYOffset);

        return containerOffset - topOffset;
    }

    // For now, we need to find the parent container because `CheckoutStep`
    // isn't the outer container yet. Once both the header and body are
    // moved inside this component, we can remove the lookup.
    private getParentContainer(): HTMLElement | undefined {
        let container: HTMLElement | null = this.containerRef.current;

        while (container && container.parentElement) {
            if (container.parentElement.classList.contains('checkout-step')) {
                return container.parentElement;
            }

            container = container.parentElement;
        }

        return this.containerRef.current ? this.containerRef.current : undefined;
    }
}
