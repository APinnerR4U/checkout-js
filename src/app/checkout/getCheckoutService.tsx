import { createCheckoutService, createLanguageService } from '@bigcommerce/checkout-sdk';

import defaultTranslations from '../language/en.json';

const languageService = createLanguageService({
    ...(window as any).language,
    defaultTranslations,
});

export default () => createCheckoutService({
    locale: languageService.getLocale(),
    shouldWarnMutation: false,
});
