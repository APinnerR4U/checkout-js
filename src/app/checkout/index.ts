import { CheckoutContextProps } from './CheckoutContext';
import { CheckoutProviderProps, CheckoutProviderState } from './CheckoutProvider';

export type CheckoutContextProps = CheckoutContextProps;
export type CheckoutProviderProps = CheckoutProviderProps;
export type CheckoutProviderState = CheckoutProviderState;

export { default as CheckoutContext } from './CheckoutContext';
export { default as CheckoutProvider } from './CheckoutProvider';
export { default as CheckoutStep } from './CheckoutStep';
export { default as withCheckout } from './withCheckout';
export { default as Checkout } from './Checkout';
export { default as CheckoutApp } from './CheckoutApp';
export { default as CheckoutSupport } from './CheckoutSupport';
export { default as NoopCheckoutSupport } from './NoopCheckoutSupport';
export { default as renderCheckout } from './renderCheckout';
