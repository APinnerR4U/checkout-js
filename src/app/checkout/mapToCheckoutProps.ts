import { CustomError } from '@bigcommerce/checkout-sdk';

import { EMPTY_ARRAY } from '../common/utility';

import getCheckoutStepStatuses from './getCheckoutStepStatuses';
import { CheckoutProps, WithCheckoutProps } from './Checkout';
import { CheckoutContextProps } from './CheckoutContext';

export default function mapToCheckoutProps(
    { checkoutService, checkoutState }: CheckoutContextProps
): WithCheckoutProps {
    const { data, errors, statuses } = checkoutState;
    const { grandTotal = 0, promotions = EMPTY_ARRAY } = data.getCheckout() || {};
    const { storeCredit = 0 } = data.getCustomer() || {};
    const submitOrderError = errors.getSubmitOrderError() as CustomError;
    const {
        checkoutSettings: { guestCheckoutEnabled: isGuestEnabled = false } = {},
        links: { loginLink: loginUrl = '' } = {},
    } = data.getConfig() || {};

    return {
        billingAddress: data.getBillingAddress(),
        cart: data.getCart(),
        clearError: checkoutService.clearError,
        consignments: data.getConsignments(),
        hasCartChanged: submitOrderError && submitOrderError.type === 'cart_changed', // TODO: Need to clear the error once it's displayed
        isGuestEnabled,
        isLoadingCheckout: statuses.isLoadingCheckout(),
        isPending: statuses.isPending(),
        loadCheckout: checkoutService.loadCheckout,
        loginUrl,
        promotions,
        steps: data.getCheckout() ? getCheckoutStepStatuses(checkoutState) : EMPTY_ARRAY,
        usableStoreCredit: Math.min(grandTotal, storeCredit),
    };
}
