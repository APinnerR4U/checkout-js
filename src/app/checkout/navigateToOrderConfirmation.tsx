import { noop } from 'lodash';

export default function navigateToOrderConfirmation(): Promise<never> {
    const url = `${window.location.href}/order-confirmation`;

    window.location.replace(url);

    return new Promise(noop);
}
