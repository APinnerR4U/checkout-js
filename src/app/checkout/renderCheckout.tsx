import React from 'react';
import ReactDOM from 'react-dom';

import CheckoutApp, { CheckoutAppProps } from './CheckoutApp';

export default function renderCheckout(
    props: CheckoutAppProps
) {
    ReactDOM.render(
        <CheckoutApp { ...props } />,
        document.getElementById(props.containerId)
    );
}
