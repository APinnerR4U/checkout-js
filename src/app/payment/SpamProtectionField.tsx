import { SpamProtectionOptions } from '@bigcommerce/checkout-sdk';
import { noop } from 'lodash';
import React, { Component } from 'react';

import { withCheckout, CheckoutContextProps } from '../checkout';

export interface SpamProtectionProps {
    containerId: string;
    onUnhandledError?(error: Error): void;
}

interface WithCheckoutSpamProtectionProps {
    initialize(options: SpamProtectionOptions): void;
}

function mapToSpamProtectionProps(
    { checkoutService }: CheckoutContextProps
): WithCheckoutSpamProtectionProps {
    return {
        initialize: checkoutService.initializeSpamProtection,
    };
}

class SpamProtectionField extends Component<SpamProtectionProps & WithCheckoutSpamProtectionProps> {
    async componentDidMount() {
        const {
            containerId,
            initialize,
            onUnhandledError = noop,
        } = this.props;

        try {
            await initialize({ containerId });
        } catch (error) {
            onUnhandledError(error);
        }
    }

    render() {
        const { containerId } = this.props;

        return <div id={ containerId }></div>;
    }
}

export default withCheckout(mapToSpamProtectionProps)(SpamProtectionField);
