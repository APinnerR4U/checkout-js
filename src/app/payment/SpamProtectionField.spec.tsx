import { createCheckoutService, CheckoutService } from '@bigcommerce/checkout-sdk';
import { mount, render } from 'enzyme';
import React, { FunctionComponent } from 'react';

import { CheckoutProvider } from '../checkout';

import SpamProtectionField, { SpamProtectionProps } from './SpamProtectionField';

describe('SpamProtectionField', () => {
    let checkoutService: CheckoutService;
    let SpamProtectionTest: FunctionComponent<SpamProtectionProps>;

    beforeEach(() => {
        checkoutService = createCheckoutService();

        jest.spyOn(checkoutService, 'initializeSpamProtection')
            .mockResolvedValue(checkoutService.getState());

        SpamProtectionTest = props => (
            <CheckoutProvider checkoutService={ checkoutService }>
                <SpamProtectionField { ...props } />
            </CheckoutProvider>
        );
    });

    it('intializes spam protection when component is mounted', () => {
        mount(<SpamProtectionTest containerId="spamProtection" />);

        expect(checkoutService.initializeSpamProtection)
            .toHaveBeenCalledWith({ containerId: 'spamProtection' });
    });

    it('renders spam protection field with the given container id', () => {
        expect(render(<SpamProtectionTest containerId="spamProtection" />))
            .toMatchSnapshot();
    });
});
