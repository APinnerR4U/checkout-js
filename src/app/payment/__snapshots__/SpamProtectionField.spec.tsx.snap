// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SpamProtectionField renders spam protection field with the given container id 1`] = `
<div
  id="spamProtection"
/>
`;
